### ISSUE TITLE
ENOSPC on linux

### ISSUE THREAD

**Issue by johanssj** [alm-tools/alm#202]
Whilst mostly a Babel user, I got interested in TypeScript a few weeks ago and started following [@basarat](https://twitter.com/basarat) on Twitter.  Seeing his tweet about Alm last night prompted me to check out Alm. In [subsequent communication, @thebabelion](https://twitter.com/thebabellion) I mentioned to [@basarat](https://twitter.com/basarat), that all did not go well and he requested I log issue here.  Here is relevant terminal output (my own path names anonymized).

In both cases *no* UI appeared.

**Terminal session in my TypeScript project directory**
`
xxx@xxx-VirtualBox ~/xxx/myproject $ cat /etc/*-release
DISTRIB_ID=LinuxMint
DISTRIB_RELEASE=18
DISTRIB_CODENAME=sarah
DISTRIB_DESCRIPTION="Linux Mint 18 Sarah"
NAME="Linux Mint"
VERSION="18 (Sarah)"
ID=linuxmint
ID_LIKE=ubuntu
PRETTY_NAME="Linux Mint 18"
VERSION_ID="18"
HOME_URL="http://www.linuxmint.com/"
SUPPORT_URL="http://forums.linuxmint.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/linuxmint/"
UBUNTU_CODENAME=xenial
cat: /etc/upstream-release: Is a directory

xxx@xxx-VirtualBox ~/xxx/myproject $ node --version
v6.6.0

xxx@xxx-VirtualBox ~/xxx/myproject $ npm --version
3.10.7

xxx@xxx-VirtualBox ~/xxx/myproject $ alm
Version: 2.0.3
DASHBOARD: http://localhost:4444
TSCONFIG: All Good! /home/xxx/xxx/myproject/tsconfig.json
[TSC] Started Initial Error Analysis: /home/xxx/xxx/myproject/tsconfig.json
[LINT] No tslint configuration found.
[TSC] Error Analysis Duration: 1s
[TSC] FileCount: 3  Errors: 2
[TSC] Incremental Error Analysis /home/xxx/xxx//myproject/tsconfig.json
[TSC] Error Analysis Duration: 1s
[TSC] FileCount: 3  Errors: 2
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: watch /home/xxx/xxx/tests/baselines/reference/project/sourceRootAbsolutePathSubfolderNoOutdir/amd/ref ENOSPC
    at exports._errnoException (util.js:1036:11)
    at FSWatcher.start (fs.js:1429:19)
    at Object.fs.watch (fs.js:1456:11)
    at createFsWatchInstance (/home/xxx/lib/node_modules/alm/node_modules/chokidar/lib/nodefs-handler.js:37:15)
    at setFsWatchListener (/home/xxx/lib/node_modules/alm/node_modules/chokidar/lib/nodefs-handler.js:80:15)
    at FSWatcher.NodeFsHandler._watchWithNodeFs (/home/xxx/lib/node_modules/alm/node_modules/chokidar/lib/nodefs-handler.js:228:14)
    at FSWatcher.NodeFsHandler._handleDir (/home/xxx/lib/node_modules/alm/node_modules/chokidar/lib/nodefs-handler.js:407:19)
    at FSWatcher.<anonymous> (/home/xxx/lib/node_modules/alm/node_modules/chokidar/lib/nodefs-handler.js:455:19)
    at FSWatcher.<anonymous> (/home/xxx/lib/node_modules/alm/node_modules/chokidar/lib/nodefs-handler.js:460:16)
    at FSReqWrap.oncomplete (fs.js:123:15)
fileListingWorker worker restarting. Don't know why it stopped with code: 1
`

**Terminal session in an empty directory**
`
xxx@xxx-VirtualBox ~ $ mkdir tmp
xxx@xxx-VirtualBox ~ $ cd tmp
xxx@xxx-VirtualBox ~/tmp $ alm
Version: 2.0.3
DASHBOARD: http://localhost:4445
[TSCONFIG]: No active project

`

Trusting this information is helpful.

Justin Johansson
[Twitter: @thebabelion](https://twitter.com/thebabellion)
[GitHub: @johanssj](https://github.com/johanssj)

**Comment by basarat** [comment]
> /home/xxx/xxx/tests/baselines/reference/project/sourceRootAbsolutePathSubfolderNoOutdir/amd/ref ENOSPC

So question : https://github.com/paulmillr/chokidar/issues/294#issuecomment-100492229 `ENOSPC means that there is no space on the drive. Is that the case?`

**Comment by basarat** [comment]
Also give `2.0.5` a go `npm install alm -g` as I just updated chokidar to see if it fixes it :rose:

**Comment by johanssj** [comment]
Just retried with 2.0.5.  Alas same result; no UI appears.  This time I noticed that when installing npm WARNs about chokidar. I don't know if this is relevant. Terminal transcripts follow.

**Install** Note WARNs at end of module tree listing.
`
> xxx@xxx-VirtualBox ~ $ npm install alm -g

/home/xxx/lib
└─┬ alm@2.0.5 
  ├── basic-auth-connect@1.0.0 
  ├── byots@2.1.0-dev.20160724.9.14 
  ├─┬ chalk@1.1.3 
  │ ├── ansi-styles@2.2.1 
  │ ├── escape-string-regexp@1.0.5 
  │ ├─┬ has-ansi@2.0.0 
  │ │ └── ansi-regex@2.0.0 
  │ ├── strip-ansi@3.0.1 
  │ └── supports-color@2.0.0 
  ├─┬ chokidar@1.6.0 
  │ ├─┬ anymatch@1.3.0 
  │ │ └─┬ micromatch@2.3.11 
  │ │   ├─┬ arr-diff@2.0.0 
  │ │   │ └── arr-flatten@1.0.1 
  │ │   ├── array-unique@0.2.1 
  │ │   ├─┬ braces@1.8.5 
  │ │   │ ├─┬ expand-range@1.8.2 
  │ │   │ │ └─┬ fill-range@2.2.3 
  │ │   │ │   ├── is-number@2.1.0 
  │ │   │ │   ├── isobject@2.1.0 
  │ │   │ │   ├── randomatic@1.1.5 
  │ │   │ │   └── repeat-string@1.5.4 
  │ │   │ ├── preserve@0.2.0 
  │ │   │ └── repeat-element@1.1.2 
  │ │   ├─┬ expand-brackets@0.1.5 
  │ │   │ └── is-posix-bracket@0.1.1 
  │ │   ├── extglob@0.3.2 
  │ │   ├── filename-regex@2.0.0 
  │ │   ├─┬ kind-of@3.0.4 
  │ │   │ └── is-buffer@1.1.4 
  │ │   ├── normalize-path@2.0.1 
  │ │   ├─┬ object.omit@2.0.0 
  │ │   │ ├─┬ for-own@0.1.4 
  │ │   │ │ └── for-in@0.1.6 
  │ │   │ └── is-extendable@0.1.1 
  │ │   ├─┬ parse-glob@3.0.4 
  │ │   │ ├── glob-base@0.3.0 
  │ │   │ └── is-dotfile@1.0.2 
  │ │   └─┬ regex-cache@0.4.3 
  │ │     ├── is-equal-shallow@0.1.3 
  │ │     └── is-primitive@2.0.0 
  │ ├── async-each@1.0.1 
  │ ├── glob-parent@2.0.0 
  │ ├── inherits@2.0.3 
  │ ├─┬ is-binary-path@1.0.1 
  │ │ └── binary-extensions@1.6.0 
  │ ├─┬ is-glob@2.0.1 
  │ │ └── is-extglob@1.0.0 
  │ ├── path-is-absolute@1.0.0 
  │ └─┬ readdirp@2.1.0 
  │   ├── graceful-fs@4.1.6 
  │   ├─┬ readable-stream@2.1.5 
  │   │ ├── buffer-shims@1.0.0 
  │   │ ├── core-util-is@1.0.2 
  │   │ ├── isarray@1.0.0 
  │   │ ├── process-nextick-args@1.0.7 
  │   │ └── string_decoder@0.10.31 
  │   └── set-immediate-shim@1.0.1 
  ├─┬ cookie-parser@1.4.3 
  │ ├── cookie@0.3.1 
  │ └── cookie-signature@1.0.6 
  ├── deep-equal@1.0.1 
  ├─┬ editorconfig@0.13.2 
  │ ├── bluebird@3.4.6 
  │ ├─┬ commander@2.9.0 
  │ │ └── graceful-readlink@1.0.1 
  │ ├─┬ lru-cache@3.2.0 
  │ │ └── pseudomap@1.0.2 
  │ └── sigmund@1.0.1 
  ├── escape-html@1.0.3 
  ├─┬ express@4.13.0 
  │ ├─┬ accepts@1.2.13 
  │ │ ├─┬ mime-types@2.1.12 
  │ │ │ └── mime-db@1.24.0 
  │ │ └── negotiator@0.5.3 
  │ ├── array-flatten@1.1.0 
  │ ├── content-disposition@0.5.0 
  │ ├── content-type@1.0.2 
  │ ├── cookie@0.1.3 
  │ ├─┬ debug@2.2.0 
  │ │ └── ms@0.7.1 
  │ ├── depd@1.0.1 
  │ ├── escape-html@1.0.2 
  │ ├── etag@1.7.0 
  │ ├─┬ finalhandler@0.4.0 
  │ │ ├── escape-html@1.0.2 
  │ │ └── unpipe@1.0.0 
  │ ├── fresh@0.3.0 
  │ ├── merge-descriptors@1.0.0 
  │ ├── methods@1.1.2 
  │ ├─┬ on-finished@2.3.0 
  │ │ └── ee-first@1.1.1 
  │ ├── parseurl@1.3.1 
  │ ├── path-to-regexp@0.1.6 
  │ ├─┬ proxy-addr@1.0.10 
  │ │ ├── forwarded@0.1.0 
  │ │ └── ipaddr.js@1.0.5 
  │ ├── qs@2.4.2 
  │ ├── range-parser@1.0.3 
  │ ├─┬ send@0.13.0 
  │ │ ├── destroy@1.0.3 
  │ │ ├── escape-html@1.0.2 
  │ │ ├── http-errors@1.3.1 
  │ │ ├── mime@1.3.4 
  │ │ └── statuses@1.2.1 
  │ ├─┬ serve-static@1.10.3 
  │ │ └─┬ send@0.13.2 
  │ │   ├── depd@1.1.0 
  │ │   └── destroy@1.0.4 
  │ ├─┬ type-is@1.6.13 
  │ │ └── media-typer@0.3.0 
  │ ├── utils-merge@1.0.0 
  │ └── vary@1.0.1 
  ├── fuzzaldrin@2.1.0 
  ├─┬ glob@7.0.6 
  │ ├── fs.realpath@1.0.0 
  │ ├─┬ inflight@1.0.5 
  │ │ └── wrappy@1.0.2 
  │ ├─┬ minimatch@3.0.3 
  │ │ └─┬ brace-expansion@1.1.6 
  │ │   ├── balanced-match@0.4.2 
  │ │   └── concat-map@0.0.1 
  │ └── once@1.4.0 
  ├── minimist@1.2.0 
  ├─┬ mkdirp@0.5.1 
  │ └── minimist@0.0.8 
  ├─┬ mocha@3.0.2 
  │ ├── browser-stdout@1.3.0 
  │ ├── diff@1.4.0 
  │ ├── glob@7.0.5 
  │ ├── growl@1.9.2 
  │ ├── json3@3.3.2 
  │ ├─┬ lodash.create@3.1.1 
  │ │ ├─┬ lodash._baseassign@3.2.0 
  │ │ │ ├── lodash._basecopy@3.0.1 
  │ │ │ └─┬ lodash.keys@3.1.2 
  │ │ │   ├── lodash._getnative@3.9.1 
  │ │ │   ├── lodash.isarguments@3.1.0 
  │ │ │   └── lodash.isarray@3.0.4 
  │ │ ├── lodash._basecreate@3.0.3 
  │ │ └── lodash._isiterateecall@3.0.9 
  │ └─┬ supports-color@3.1.2 
  │   └── has-flag@1.0.0 
  ├── monaco@1.201608152240.0 
  ├─┬ monaco-css@1.1.2 
  │ ├─┬ vscode-css-languageservice@1.0.5 
  │ │ └── vscode-nls@1.0.7 
  │ └── vscode-languageserver-types@1.0.3 
  ├── monaco-languages@0.3.1 
  ├─┬ multimatch@2.1.0 
  │ ├── array-differ@1.0.0 
  │ ├─┬ array-union@1.0.2 
  │ │ └── array-uniq@1.0.3 
  │ └── arrify@1.0.1 
  ├─┬ mv@2.1.1 
  │ └─┬ rimraf@2.4.5 
  │   └── glob@6.0.4 
  ├── ncp@2.0.0 
  ├─┬ node-fetch@1.6.1 
  │ ├─┬ encoding@0.1.12 
  │ │ └── iconv-lite@0.4.13 
  │ └── is-stream@1.1.0 
  ├── open@0.0.5 
  ├─┬ ora@0.2.3 
  │ ├─┬ cli-cursor@1.0.2 
  │ │ └─┬ restore-cursor@1.0.1 
  │ │   ├── exit-hook@1.1.1 
  │ │   └── onetime@1.1.0 
  │ ├── cli-spinners@0.1.2 
  │ └── object-assign@4.1.0 
  ├── rimraf@2.5.4 
  ├─┬ socket.io@1.4.8 
  │ ├─┬ engine.io@1.6.11 
  │ │ ├─┬ accepts@1.1.4 
  │ │ │ ├─┬ mime-types@2.0.14 
  │ │ │ │ └── mime-db@1.12.0 
  │ │ │ └── negotiator@0.4.9 
  │ │ ├── base64id@0.1.0 
  │ │ ├─┬ engine.io-parser@1.2.4 
  │ │ │ ├── after@0.8.1 
  │ │ │ ├── arraybuffer.slice@0.0.6 
  │ │ │ ├── base64-arraybuffer@0.1.2 
  │ │ │ ├── blob@0.0.4 
  │ │ │ ├─┬ has-binary@0.1.6 
  │ │ │ │ └── isarray@0.0.1 
  │ │ │ └── utf8@2.1.0 
  │ │ └─┬ ws@1.1.0 
  │ │   ├── options@0.0.6 
  │ │   └── ultron@1.0.2 
  │ ├─┬ has-binary@0.1.7 
  │ │ └── isarray@0.0.1 
  │ ├─┬ socket.io-adapter@0.4.0 
  │ │ └─┬ socket.io-parser@2.2.2 
  │ │   ├── debug@0.7.4 
  │ │   ├── isarray@0.0.1 
  │ │   └── json3@3.2.6 
  │ ├─┬ socket.io-client@1.4.8 
  │ │ ├── backo2@1.0.2 
  │ │ ├── component-bind@1.0.0 
  │ │ ├── component-emitter@1.2.0 
  │ │ ├─┬ engine.io-client@1.6.11 
  │ │ │ ├── component-inherit@0.0.3 
  │ │ │ ├── has-cors@1.1.0 
  │ │ │ ├── parsejson@0.0.1 
  │ │ │ ├── parseqs@0.0.2 
  │ │ │ ├── ws@1.0.1 
  │ │ │ ├── xmlhttprequest-ssl@1.5.1 
  │ │ │ └── yeast@0.1.2 
  │ │ ├── indexof@0.0.1 
  │ │ ├── object-component@0.0.3 
  │ │ ├─┬ parseuri@0.0.4 
  │ │ │ └─┬ better-assert@1.0.2 
  │ │ │   └── callsite@1.0.0 
  │ │ └── to-array@0.1.4 
  │ └─┬ socket.io-parser@2.2.6 
  │   ├── benchmark@1.0.0 
  │   ├── component-emitter@1.1.2 
  │   └── isarray@0.0.1 
  ├─┬ ts-node@1.3.0 
  │ ├── diff@2.2.3 
  │ ├── make-error@1.2.1 
  │ ├── pinkie@2.0.4 
  │ ├─┬ source-map-support@0.4.2 
  │ │ └─┬ source-map@0.1.32 
  │ │   └── amdefine@1.0.0 
  │ ├─┬ tsconfig@5.0.3 
  │ │ ├── any-promise@1.3.0 
  │ │ ├─┬ parse-json@2.2.0 
  │ │ │ └─┬ error-ex@1.3.0 
  │ │ │   └── is-arrayish@0.2.1 
  │ │ ├─┬ strip-bom@2.0.0 
  │ │ │ └── is-utf8@0.2.1 
  │ │ └── strip-json-comments@2.0.1 
  │ └── xtend@4.0.1 
  ├─┬ tslint@3.15.0-dev.0 
  │ ├── colors@1.1.2 
  │ ├── diff@2.2.3 
  │ ├─┬ findup-sync@0.3.0 
  │ │ └── glob@5.0.15 
  │ ├─┬ optimist@0.6.1 
  │ │ ├── minimist@0.0.10 
  │ │ └── wordwrap@0.0.3 
  │ ├── resolve@1.1.7 
  │ └─┬ underscore.string@3.3.4 
  │   ├── sprintf-js@1.0.3 
  │   └── util-deprecate@1.0.2 
  ├── typescript@2.1.0-dev.20160912 
  └─┬ update-notifier@1.0.2 
    ├─┬ boxen@0.6.0 
    │ ├── ansi-align@1.1.0 
    │ ├── camelcase@2.1.1 
    │ ├── cli-boxes@1.0.0 
    │ ├── filled-array@1.1.0 
    │ ├─┬ repeating@2.0.1 
    │ │ └─┬ is-finite@1.0.1 
    │ │   └── number-is-nan@1.0.0 
    │ ├─┬ string-width@1.0.2 
    │ │ ├── code-point-at@1.0.0 
    │ │ └── is-fullwidth-code-point@1.0.0 
    │ └── widest-line@1.0.0 
    ├─┬ configstore@2.1.0 
    │ ├─┬ dot-prop@3.0.0 
    │ │ └── is-obj@1.0.1 
    │ ├── os-tmpdir@1.0.1 
    │ ├── osenv@0.1.3 
    │ ├── uuid@2.0.3 
    │ └─┬ write-file-atomic@1.2.0 
    │   ├── imurmurhash@0.1.4 
    │   └── slide@1.1.6 
    ├── is-npm@1.0.0 
    ├─┬ latest-version@2.0.0 
    │ └─┬ package-json@2.4.0 
    │   ├─┬ got@5.6.0 
    │   │ ├─┬ create-error-class@3.0.2 
    │   │ │ └── capture-stack-trace@1.0.0 
    │   │ ├── duplexer2@0.1.4 
    │   │ ├── is-plain-obj@1.1.0 
    │   │ ├── is-redirect@1.0.0 
    │   │ ├── is-retry-allowed@1.1.0 
    │   │ ├── lowercase-keys@1.0.0 
    │   │ ├── node-status-codes@1.0.0 
    │   │ ├── pinkie-promise@2.0.1 
    │   │ ├── read-all-stream@3.1.0 
    │   │ ├── timed-out@2.0.0 
    │   │ ├── unzip-response@1.0.1 
    │   │ └─┬ url-parse-lax@1.0.0 
    │   │   └── prepend-http@1.0.4 
    │   ├─┬ registry-auth-token@3.0.1 
    │   │ └─┬ rc@1.1.6 
    │   │   ├── deep-extend@0.4.1 
    │   │   ├── ini@1.3.4 
    │   │   └── strip-json-comments@1.0.4 
    │   └── registry-url@3.1.0 
    ├── lazy-req@1.1.0 
    ├─┬ semver-diff@2.1.0 
    │ └── semver@5.3.0 
    └─┬ xdg-basedir@2.0.0 
      └── os-homedir@1.0.1 

npm WARN optional Skipping failed optional dependency /alm/chokidar/fsevents:
npm WARN notsup Not compatible with your operating system or architecture: fsevents@1.0.14
`

**Now in project directory**

Observation: While my current directory in this instance is ~/workspace/myproject , it seems strange that alm is looking at a sibling directory ~/workspace/TypeScript/ in the trace below where it says 'Error: watch'.  Strikes me as odd that it would do that.

`
> xxx@xxx-VirtualBox ~/workspace/myproject $ alm
Version: 2.0.5
DASHBOARD: http://localhost:4444
TSCONFIG: All Good! /home/xxx/workspace/myproject/tsconfig.json
[TSC] Started Initial Error Analysis: /home/xxx/workspace/myproject/tsconfig.json
[LINT] No tslint configuration found.
[TSC] Error Analysis Duration: 1s
[TSC] FileCount: 3  Errors: 2
[TSC] Incremental Error Analysis /home/xxx/workspace/myproject/tsconfig.json
[TSC] Error Analysis Duration: 1s
[TSC] FileCount: 3  Errors: 2
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: watch /home/xxx/workspace/TypeScript/tests/baselines/reference/project/sourceRootAbsolutePathModuleSubfolderSpecifyOutputDirectory/amd/outdir ENOSPC
    at exports._errnoException (util.js:1036:11)
    at FSWatcher.start (fs.js:1429:19)
    at Object.fs.watch (fs.js:1456:11)
    at createFsWatchInstance (/home/xxx/lib/node_modules/alm/node_modules/chokidar/lib/nodefs-handler.js:37:15)
    at setFsWatchListener (/home/xxx/lib/node_modules/alm/node_modules/chokidar/lib/nodefs-handler.js:80:15)
    at FSWatcher.NodeFsHandler._watchWithNodeFs (/home/xxx/lib/node_modules/alm/node_modules/chokidar/lib/nodefs-handler.js:228:14)
    at FSWatcher.NodeFsHandler._handleDir (/home/xxx/lib/node_modules/alm/node_modules/chokidar/lib/nodefs-handler.js:407:19)
    at FSWatcher.<anonymous> (/home/xxx/lib/node_modules/alm/node_modules/chokidar/lib/nodefs-handler.js:455:19)
    at FSWatcher.<anonymous> (/home/xxx/lib/node_modules/alm/node_modules/chokidar/lib/nodefs-handler.js:460:16)
    at FSReqWrap.oncomplete (fs.js:123:15)
fileListingWorker worker restarting. Don't know why it stopped with code: 1
`

**Now in empty directory**
`
> xxx@xxx-VirtualBox ~/tmp $ alm
Version: 2.0.5
DASHBOARD: http://localhost:4444
[TSCONFIG]: No active project
`

Again no UI appears; just stares me back. :-(

**Comment by basarat** [comment]
> Now in empty directory

That one is actually *expected*.  You need to open the dashboard URL `DASHBOARD: http://localhost:4444` in chrome yourself. Or if you want you can run `alm -o` which will open chrome for you. Its a nodejs web app in the true sense i.e nodejs backend and chrome frontend. Sorry about any confusion there but its menioned in the usage : https://github.com/alm-tools/alm#usage and the quickstart : https://basarat.gitbooks.io/alm/content/quickstart.html 

However why the `ENOSPC` happens in the first case is still not understood. That said, I think the UI is functional in that case too (you just need to open it ... or run `alm -o`) :rose:

**Comment by johanssj** [comment]
Yeah, sorry, I just realised that. My expectation was that some UI would happen just like when you launch Atom editor from the command line. Mia culpa for not reading the docs; but then again one cannot assume people will read the docs so at least a clear message in the terminal output on start up might be a good idea. (Yes, it did say DASHBOARD: http://etc, but you know.)  At least I can get sort of going with it now.

**Comment by basarat** [comment]
> then again one cannot assume people will read the doc

Agreed :rose: Fixed and released with `2.0.6` :heart: 

![image](https://cloud.githubusercontent.com/assets/874898/18622477/f4d10fa6-7e72-11e6-9394-7b47818f4971.png)

(Screenshot shows `2.0.5` as that is the version I had while adding this fix).

**Comment by johanssj** [comment]
Good on you. While I feel like a goose, I don't really think I'm a goose. Thinking about the psychology of it though, probably had I run it the empty directory scenario first up and didn't see all the stack trace messages, I would have opened up the browser straight away.  Still like you allude to, the ENOSPC issue in the first case needs understanding as to why.

**Comment by basarat** [comment]
> Still like you allude to, the ENOSPC issue in the first case needs understanding as to why.

Editing the issue to track that better. This is external stuff but googling around the following shows up:

* Hows the space in `/tmp`: http://stackoverflow.com/a/22476114/390330 
* Seems to be helping a lot of people (upvotes) : http://stackoverflow.com/a/32600959/390330 

Just as an estimate how many files are in that project folder? :rose:

**Comment by johanssj** [comment]
There were less than a dozen of mine but maybe it was looking into 
node_modules etc which could be many.

I will do some further tests/research. Might take a day or so as have 
other priorities. Will try to squeeze in asap.

**Comment by johanssj** [comment]
Got a better handle on the problem now.
A bit tedious & probably a bit specific to my particular machine capacity but nevertheless ...
We are in directory ~/tmp which has nothing else but 2 alm torture-test scripts in it.

```
  > ~/tmp $ ls
  alm-torture-test-fail.sh  alm-torture-test-pass.sh
```

These scripts are:

alm-torture-test-pass.sh:
```
mkdir workspace-september
cd workspace-september
git clone https://github.com/DefinitelyTyped/DefinitelyTyped.git
cp -R DefinitelyTyped DefinitelyTyped2
cp -R DefinitelyTyped DefinitelyTyped3
cp -R DefinitelyTyped DefinitelyTyped4
cp -R DefinitelyTyped DefinitelyTyped5
cp -R DefinitelyTyped DefinitelyTyped6
cp -R DefinitelyTyped DefinitelyTyped7
cp -R DefinitelyTyped DefinitelyTyped8
cp -R DefinitelyTyped DefinitelyTyped9
mkdir empty
cd empty
alm
```

and

alm-torture-test-fail.sh:  *(Notice the workspace direcory name now has parens around the abbreviated month name)*
```
rm -rf workspace-september
mkdir "workspace(sep)"
cd "workspace(sep)"
git clone https://github.com/DefinitelyTyped/DefinitelyTyped.git
cp -R DefinitelyTyped DefinitelyTyped2
cp -R DefinitelyTyped DefinitelyTyped3
cp -R DefinitelyTyped DefinitelyTyped4
cp -R DefinitelyTyped DefinitelyTyped5
mkdir empty
cd empty
alm
```

**Running alm-torture-test-pass.sh** as the name suggests is no problem.

In a workspace subdirectory, it clones a large repo off GitHub, makes 8 copies of the repo, then creates and changes into an empty directory, and then in the empty directory runs alm.  All without any issues.

```
> ~/tmp $ ./alm-torture-test-pass.sh
Cloning into 'DefinitelyTyped'...
remote: Counting objects: 130601, done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 130601 (delta 12), reused 12 (delta 12), pack-reused 130586
Receiving objects: 100% (130601/130601), 121.07 MiB | 238.00 KiB/s, done.
Resolving deltas: 100% (76234/76234), done.
Checking connectivity... done.
Version: 2.0.7
DASHBOARD: (Please open in chrome) http://localhost:4444
[TSCONFIG]: No active project
```

**Now running alm-torture-test-fail.sh** as the name suggests *is a problem* (at least on my machine).

Again, in a workspace subdirectory, it clones the same large repo off GitHub,
makes 4 copies of the repo (4 less than in the previous test), then creates and changes into an empty directory, and then in the empty directory runs alm.  This time ENOSPC shows up.

```
> ~/tmp $ ./alm-torture-test-fail.sh
Cloning into 'DefinitelyTyped'...
remote: Counting objects: 130601, done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 130601 (delta 12), reused 12 (delta 12), pack-reused 130586
Receiving objects: 100% (130601/130601), 121.07 MiB | 238.00 KiB/s, done.
Resolving deltas: 100% (76234/76234), done.
Checking connectivity... done.
Version: 2.0.7
DASHBOARD: (Please open in chrome) http://localhost:4444
[TSCONFIG]: No active project
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: watch /home/mars/tmp/workspace(sep)/DefinitelyTyped4/sencha_touch ENOSPC
    at exports._errnoException (util.js:1036:11)
    at FSWatcher.start (fs.js:1429:19)
    at Object.fs.watch (fs.js:1456:11)
    at createFsWatchInstance (/home/mars/lib/node_modules/alm/node_modules/chokidar/lib/nodefs-handler.js:37:15)
    at setFsWatchListener (/home/mars/lib/node_modules/alm/node_modules/chokidar/lib/nodefs-handler.js:80:15)
    at FSWatcher.NodeFsHandler._watchWithNodeFs (/home/mars/lib/node_modules/alm/node_modules/chokidar/lib/nodefs-handler.js:228:14)
    at FSWatcher.NodeFsHandler._handleDir (/home/mars/lib/node_modules/alm/node_modules/chokidar/lib/nodefs-handler.js:407:19)
    at FSWatcher.<anonymous> (/home/mars/lib/node_modules/alm/node_modules/chokidar/lib/nodefs-handler.js:455:19)
    at FSWatcher.<anonymous> (/home/mars/lib/node_modules/alm/node_modules/chokidar/lib/nodefs-handler.js:460:16)
    at FSReqWrap.oncomplete (fs.js:123:15)
fileListingWorker worker restarting. Don't know why it stopped with code: 1
```

**Further**
When the number of copies of the repo are reduced down to two, so we have DefinitelyTyped2 and DefinitelyTyped3, the problem goes away.

**Conclusion**
The problem manifests when there are a lot of files in sibling directories of the empty directory in which alm is run AND when a parent directory of that (empty) directory has paren characters in its name.

So now some questions, like
- why didn't the -pass script fail when there were twice of many files about?
- what do parens in directory names have to do with it?
- even with parens in directory names why does problem go away with fewer files about?
- what business of alm or it's file watcher is it to be peeking into sibling directories of the current project directory in any case.

btw. I organise my workspaces by years with 12 subdirs being each month in respective year.  A symlink in my home directory called current-workspace always points to the latest month. Then because I like my workspace directory listings to appear nicely ordered, I call name individual monthly workspaces like workspace-2016-03(mar) because then I don't have to think what name of month is month 03.

I think I have spent too much time on this now but will be pleased if this information is useful to you.

Cheers
Justin Johansson

Maybe see you on Twitter, [thebabellion](https://twitter.com/thebabellion)

**Comment by basarat** [comment]
I've also added docs : https://basarat.gitbooks.io/alm/content/contributing/file-listing.html to help you and others.

> what business of alm or it's file watcher is it to be peeking into sibling directories of the current project directory in any case

Sounds very weird. FWIW our external API access to the watcher is super simple : https://github.com/alm-tools/alm/blob/13ceedbb2c0981423da28ed11821f87baac206f4/src/server/workers/fileListing/fileListingWorker.ts#L252 fingers crossed that chokidar fixes it or someone reports it :rose:

**Comment by johanssj** [comment]
I was thinking of bowing out of this issue for the moment but feel it is 
too important to let slip.

Certainly it would be great if "chokidar fixes it or someone [else] 
reports it".

If you keep this issue open for a while I'll try to get back to it in a 
few days. Might have to fork chokidar & see what's cooking.

**Comment by johanssj** [comment]
btw. Were you able to replicate my test scripts with opening alm in the 
empty directory with all those copies of DefinitelyTyped as siblings, or 
did you just take my terminal output as "proof"? Admittedly the tests 
were a bit tedious so I understand if you couldn't afford the time to do so.

**Comment by basarat** [comment]
> Admittedly the tests
were a bit tedious so I understand if you couldn't afford the time to do so.

Haven't tried it 😁

**Comment by johanssj** [comment]
I've had the opportunity to dig deep into this issue having forked both alm and chokidar. Also read up on the Node FS watch API. I feel that these issues will never go away with the current watch architecture as the low-level watch API is intrinsically flakey.

Even with Eclipse IDE which I have configured to use native file system hooks, I find that sometimes my workspace does not get refreshed when it is supposed to.

Having spent over 5 years of my career writing file system drivers from encryption to security applications I think I can speak with qualified experience.

Notice that your profile says you live in Melbourne.  Long time ago I was lead engineer for an anti-virus/PC security company in Melbourne. These days I'm back in Adelaide .. an easy phone call away if ever you wish to discuss.  Talking is probably a more effective means of communicating than typing too many words.  Anyway you are welcome to the offer if you wish. Contact via DM on Twitter, [thebabellion](https://twitter.com/thebabellion).

**Comment by basarat** [comment]
> Having spent over 5 years of my career writing file system drivers from encryption to security applications I think I can speak with qualified experience

Agree 100%. That's why offloaded to chokidar as it is good enough for most projects :rose:

**Comment by johanssj** [comment]
Unfortunately though I will not be able to use alm editor. I put some 
console.log messages into alm file watch callbacks. Sometimes they are 
called when files change sometimes (mostly) no.  Reboot (LinuxMint) and 
I'll see those hooks being called again for a while.  Copy some files 
around, do a git clone in a sibling directory (to where alm was 
launched) and then flop again, no more callbacks. Just dead in the water.

**Comment by ishigami** [comment]
About the ENOSPC error on Linux, I've tried the following code and fixed the error

`echo fs.inotify.max_user_watches=524288 | sudo tee -a /etc/sysctl.conf && sudo sysctl -p`

More info at: https://github.com/typicode/json-server/issues/361

**Comment by johanssj** [comment]
Thanks for this.

I already had placed

|fs.inotify.max_user_watches=524288|

at the end of the file AND restarting the box

|/etc/sysctl.conf|

but do not recall ever issuing the command

|sudo sysctl -p
|
So in your opinion that should make a difference?

**Comment by johanssj** [comment]
Just saw some emails from issues in paulmillr/chokidar which say 
somethings that sound relevant to what I reported on this ENOSPC problem 
a few weeks ago. It is encouraging that others have reported these 
things & that perhaps some fixes might be happening in chokidar.

"Any news on this? Also a default-off option is welcome. I see there's 
already a PR, but it was not accepted. Atm I can't use |livereload| 
because I'm working in a folder containing parenthesis..."

and

"Problems with parentheses will be solved by #535 
<https://github.com/paulmillr/chokidar/pull/535> once the apparent 
windows issues there are resolved.

A different fix will still be needed for other sorts of glob characters 
like asterisks."

I do remember mentioning the parentheses thing as well .. so I it 
definitely was not my imagination afterall.

**Comment by basarat** [comment]
Closing as external and can't think of a fix short of reinventing chokidar which is not on the roadmap at the moment.
